### PR TITLE
Terraform Cloudwatch to Firehose Aux Template

### PR DESCRIPTION
**Pull request by horakj**
## Background

Customer requested a way to ingest Cloudwatch logs into Panther for real time alerts utilizing Terraform.

## Changes

- Created Terraform AWS template for that connects Cloudwatch subscriptions > Kinesis Firehose > S3 
- Files: Main.tf and variables.tf

## Testing

- Terraform Validate

### CODE REVIEW

File: deployments/auxiliary/terraform/cloudwatch_firehose/main.tf [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+
+# Sample template for gathering Cloudwatch logs into S3 via Firehose.
+
+
+### Cloudwatch Subsciption
+resource "aws_cloudwatch_log_subscription_filter" "cloudwatch_log_filter" {
+  name            = "cloudwatch_log_filter"
+  role_arn        = aws_iam_role.cloudwatch_firehose_profile.arn
+  log_group_name  = var.log_group_name
+  filter_pattern  = "logtype cloudwatch"
```
Comment (jacknagz) [inline deployments/auxiliary/terraform/cloudwatch_firehose/main.tf:34]: Interesting - what's the reason for this filter pattern?

File: deployments/auxiliary/terraform/cloudwatch_firehose/main.tf [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+  destination = "extended_s3"
+
+  extended_s3_configuration {
+    bucket_arn         = aws_s3_bucket.cloudwatch_data_bucket.arn
+    role_arn           = aws_iam_role.cloudwatch_data_firehose_role.arn
+    prefix             = "cloudwatchlogs/"
+    compression_format = "GZIP"
+
+    # Data is flushed once one of the buffer hints are satisfied
+    buffer_interval = 300
```
Comment (jacknagz) [inline deployments/auxiliary/terraform/cloudwatch_firehose/main.tf:109]: These should be variables IMO

Comment (jacknagz) [reply deployments/auxiliary/terraform/cloudwatch_firehose/main.tf:109]: @rleighton @kostaspap what do you think are reasonable settings here? I used to just max them both out, but users might want lower latency

File: deployments/auxiliary/terraform/cloudwatch_firehose/main.tf [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+    }
+  }
+
+  # Short expiration because this data is sent to Panther.
+  # This can be adjusted per your individual needs.
+  lifecycle_rule {
+    id      = "30DayExpiration"
+    enabled = true
+    expiration {
+      days = 30
```
Comment (jacknagz) [inline deployments/auxiliary/terraform/cloudwatch_firehose/main.tf:186]: I'd also make this a variable

**Review by jacknagz — APPROVED** [review]
Great job on this!

File: deployments/auxiliary/terraform/cloudwatch_firehose/main.tf (hunk 1 above)
Comment (horakj) [reply deployments/auxiliary/terraform/cloudwatch_firehose/main.tf:34]: filter_pattern - (Required) A valid CloudWatch Logs filter pattern for subscribing to a filtered stream of log events.

https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/cloudwatch_log_subscription_filter

File: deployments/auxiliary/terraform/cloudwatch_firehose/main.tf (hunk 2 above)
Comment (austinbyers) [reply deployments/auxiliary/terraform/cloudwatch_firehose/main.tf:109]: Agreed - make them variables (but they can have reasonable default values)

File: deployments/auxiliary/terraform/cloudwatch_firehose/main.tf (hunk 1 above)
Comment (jacknagz) [reply deployments/auxiliary/terraform/cloudwatch_firehose/main.tf:34]: Does this work end-to-end? I feel like that filter pattern wouldn't match anything

File: deployments/auxiliary/terraform/cloudwatch_firehose/main.tf [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+# Copyright (C) 2020 Panther Labs Inc
+#
+# Panther Enterprise is licensed under the terms of a commercial license available from
+# Panther Labs Inc ("Panther Commercial License") by contacting contact@runpanther.com.
+# All use, distribution, and/or modification of this software, whether commercial or non-commercial,
+# falls under the Panther Commercial License to the extent it is permitted.
```
Comment (austinbyers) [inline deployments/auxiliary/terraform/cloudwatch_firehose/main.tf:22]: Delete the extra license header

File: deployments/auxiliary/terraform/cloudwatch_firehose/main.tf [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+
+
+
+# Sample template for gathering Cloudwatch logs into S3 via Firehose.
+
+
+### Cloudwatch Subsciption
+resource "aws_cloudwatch_log_subscription_filter" "cloudwatch_log_filter" {
+  name            = "cloudwatch_log_filter"
```
Comment (austinbyers) [inline deployments/auxiliary/terraform/cloudwatch_firehose/main.tf:31]: I assume this is a resource name that has to be unique within an account/region? In which case, should it be something like "panther_cloudwatch_ingestion"?

File: deployments/auxiliary/terraform/cloudwatch_firehose/main.tf [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+resource "aws_s3_bucket_notification" "cloudwatch_data_bucket" {
+  bucket = aws_s3_bucket.cloudwatch_data_bucket.id
+
+  topic {
+    events = [
+      "s3:ObjectCreated:*"
+    ]
+    topic_arn = var.s3_notifications_topic
+  }
+}
```
Comment (austinbyers) [inline deployments/auxiliary/terraform/cloudwatch_firehose/main.tf:231]: Add newline (unless the formatting removes it automatically)

File: deployments/auxiliary/terraform/cloudwatch_firehose/variables.tf [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+# Copyright (C) 2020 Panther Labs Inc
+#
+# Panther Enterprise is licensed under the terms of a commercial license available from
+# Panther Labs Inc ("Panther Commercial License") by contacting contact@runpanther.com.
+# All use, distribution, and/or modification of this software, whether commercial or non-commercial,
+# falls under the Panther Commercial License to the extent it is permitted.
```
Comment (austinbyers) [inline deployments/auxiliary/terraform/cloudwatch_firehose/variables.tf:22]: Remove extra license header

File: deployments/auxiliary/terraform/cloudwatch_firehose/variables.tf [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+}
+
+variable "log_group_name" {
+  type        = string
+  description = "The name of the log group to associate the subscription filter with"
+}
+
+variable "s3_notifications_topic" {
+  type        = string
+  description = "SNS topic arn to send S3 notifications for pulling data"
```
Comment (austinbyers) [inline deployments/auxiliary/terraform/cloudwatch_firehose/variables.tf:42]: Is this topic typically created by a Panther deployment? If so, please note that. Make it clear which resources are in the customer ("auxiliary") account and which would be in Panther

**Review by austinbyers — COMMENTED** [review]
Nice work!

File: deployments/auxiliary/terraform/cloudwatch_firehose/variables.tf (hunk 8 above)
Comment (horakj) [reply deployments/auxiliary/terraform/cloudwatch_firehose/variables.tf:42]: This is created with the S3 source setup cloudformation

File: deployments/auxiliary/terraform/cloudwatch_firehose/main.tf (hunk 1 above)
Comment (horakj) [reply deployments/auxiliary/terraform/cloudwatch_firehose/main.tf:34]: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-logs-subscriptionfilter.html

Filter Pattern is just a string to filter on

Comment (jacknagz) [reply deployments/auxiliary/terraform/cloudwatch_firehose/main.tf:34]: Right, it's an expression that determines if logs are ultimately sent or not

File: deployments/auxiliary/terraform/cloudwatch_firehose/main.tf (hunk 4 above)
Comment (horakj) [reply deployments/auxiliary/terraform/cloudwatch_firehose/main.tf:22]: For some reason there is a duplicate that was created but not by me but something on here. When I make the next commit I will see if this continues to be an issue and adjust.

**Comment by austinbyers** [comment]
@horakj this has been open for over a month now - any chance we can get it over the finish line?

**Comment by horakj** [comment]
@austinbyers I am closing this PR and will come back to it at a later time. It is low priority at this moment.

**Pull request closed without merging**